streams/src/test/java/org/apache/kafka/streams/integration/EosBetaUpgradeIntegrationTest.java

```diff
@@ -0,0 +1,691 @@
+    private static final int MAX_POLL_INTERVAL_MS = 5 * 1000;
+    private static final int MAX_WAIT_TIME_MS = 60 * 1000;
+
+    @ClassRule
+    public static final EmbeddedKafkaCluster CLUSTER = new EmbeddedKafkaCluster(
+        NUM_BROKERS,
+        Utils.mkProperties(Collections.singletonMap("auto.create.topics.enable", "false"))
+    );
+
+    private static String applicationId;
+    private final static int NUM_TOPIC_PARTITIONS = 2;
+    private final static String CONSUMER_GROUP_ID = "readCommitted";
+    private final static String MULTI_PARTITION_INPUT_TOPIC = "multiPartitionInputTopic";
+    private final static String MULTI_PARTITION_OUTPUT_TOPIC = "multiPartitionOutputTopic";
+    private final String storeName = "store";
+
+    private AtomicBoolean errorInjected;
+    private AtomicInteger commitRequested;
+    private Throwable uncaughtException;
+
+    private int testNumber = 0;
+
+    @Before
+    public void createTopics() throws Exception {
+        applicationId = "appId-" + ++testNumber;
+        CLUSTER.deleteTopicsAndWait(
+            MULTI_PARTITION_INPUT_TOPIC,
+            MULTI_PARTITION_OUTPUT_TOPIC,
+            applicationId + "-" + storeName + "-changelog"
+        );
+
+        CLUSTER.createTopic(MULTI_PARTITION_INPUT_TOPIC, NUM_TOPIC_PARTITIONS, 1);
+        CLUSTER.createTopic(MULTI_PARTITION_OUTPUT_TOPIC, NUM_TOPIC_PARTITIONS, 1);
+    }
+
+    @Test
+    public void shouldUpgradeFromEosAlphaToEosBeta() throws Exception {
+        // We use two KafkaStreams clients that we upgrade from eos-alpha to eos-beta. During the upgrade,
+        // we ensure that there are pending transaction and verify that data is processed correctly.
+        //
+        // We either close clients cleanly (`injectError = false`) or let them crash (`injectError = true`) during
+        // the upgrade. For both cases, EOS should not be violated.
+        //
+        // The commit interval is set to MAX_VALUE and the used `Processor` request commits manually so we have full
+        // control when a commit actually happens. We use an input topic with 2 partitions and each task will request
+        // a commit after processing 10 records.
+        //
+        // 1.  start both clients and wait until rebalance stabilizes
+        // 2.  write 10 records per input topic partition and verify that the result was committed
+        // 3.  write 5 records per input topic partition to get pending transactions (verified via "read_uncommitted" mode)
+        //      - both pending transactions are based on task producers
+        // 4.  stop/crash the first client, wait until rebalance stabilizes:
+        //      - stop case:
+        //        * verify that the stopped client did commit its pending transaction during shutdown
+        //        * the second client will still have an pending transaction
+        //      - crash case:
+        //        * the second client will have two pending transactions
```

Comment:
The second client is the "healthy" one: it will host one task and has one open transaction before the crash. The first client also hosts one task before the crash. When the first client fails, it will abort its transaction (with zero pending output records written) during shutdown and its task will migrate to the second client. The second client won't "touch" its original task and only create a second task, start a new TX for the new task (it would call init and would abort a pending TX if there would be any, but is also starts it's _own_ new TX) and process 5 records without committing the TX because it would only commit after 10 records. Hence, it "stabilizes" with 2 pending transactions with 5 writes each.

Does this make sense?

Reply:
I thought the first client would only crash after it has processed 5 records and hence there's an ongoing transaction already?

Reply:
Not in the current way the test is written...